### ISSUE TITLE
GWT should support "Content Security Policy"

### ISSUE THREAD

**Issue by dankurka** [gwtproject/gwt#8197]
Originally reported on Google Code with ID 8207

```
Found in GWT Release   : any
Encountered on OS      : any OS
Encountered on Browser : FF, Chrome

Detailed description (please be as specific as possible):

GWT should support CSP

https://dvcs.w3.org/hg/content-security-policy/raw-file/tip/csp-specification.dev.html

Current issues:
- Devmode uses eval, which is forbidden by CSP
- History support with iFrame is not accepted by CSP

Used CSP HTTP Header:
Content-Security-Policy:   default-src 'self' gwt.google.com; img-src 'self'
X-Content-Security-Policy: default-src 'self' gwt.google.com; img-src 'self'

GWT History iFrame in Hostpage:
<iframe src="javascript:''" id="__gwt_historyFrame" style="width:0;height:0;border:0"></iframe>

Exception:
08:30:06.117 [ERROR] [csptest] Unable to load module entry point class null (see associated
exception for details)

com.google.gwt.core.client.JavaScriptException: (Error) @com.google.gwt.core.client.impl.Impl::registerEntry()([]):
call to eval() blocked by CSP
    at com.google.gwt.dev.shell.BrowserChannelServer.invokeJavascript(BrowserChannelServer.java:249)
    at com.google.gwt.dev.shell.ModuleSpaceOOPHM.doInvoke(ModuleSpaceOOPHM.java:136)
    at com.google.gwt.dev.shell.ModuleSpace.invokeNative(ModuleSpace.java:571)
    at com.google.gwt.dev.shell.ModuleSpace.invokeNativeObject(ModuleSpace.java:279)
    at com.google.gwt.dev.shell.JavaScriptHost.invokeNativeObject(JavaScriptHost.java:91)
    at com.google.gwt.core.client.impl.Impl.registerEntry(Impl.java)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at com.google.gwt.dev.shell.ModuleSpace.onLoad(ModuleSpace.java:357)
    at com.google.gwt.dev.shell.OophmSessionHandler.loadModule(OophmSessionHandler.java:200)
    at com.google.gwt.dev.shell.BrowserChannelServer.processConnection(BrowserChannelServer.java:526)
    at com.google.gwt.dev.shell.BrowserChannelServer.run(BrowserChannelServer.java:364)
    at java.lang.Thread.run(Thread.java:722)

Firebug Console Output:
CSP WARN: Directive inline script base restriction violated
javascript:''
CSP WARN: Directive eval script base restriction violated
call to eval() or related function blocked by CSP


Workaround if you have one: not using CSP
```

Reported by `daniel.gerber2` on 2013-06-19 06:50:37

**Comment by akbertram** [comment]
Has this been released as part of 2.7.0? According to the linked code review the patch seems to have been merged?

**Comment by akbertram** [comment]
Oh actually this look like it has been imported from much earlier work. Is there any reason it is still open?

**Comment by tbroyer** [comment]
The linked review has been merged and is in 2.8.0-beta1 (but not 2.7.0). There are still a few widgets making use of `src="javascript:''"`  though (namely `FormPanel`, `NamedFrame`, and `RichTextArea` –in IE8/9/10 only though), or `href="javascript:''"` or similar (namely `CellTree` –for its “load more” link'y false-button–, `Anchor` –if its href is left empty–, and `DisclosurePanel`).
I'm hesitant to close this issue given those, and the fact you need to use the `direct_install` linker instead of the (now-default) `xsiframe` linker. @dankurka ?

**Comment by tbroyer** [comment]
`FormPanel` just got updated to use `href="about:blank"`: https://gwt-review.googlesource.com/17740

**Comment by nandakishoremutyala** [comment]
Do we have CSP support for GWT in latest version (2.7/2.8). Please confirm?

**Comment by tbroyer** [comment]
@nandakishoremutyala Depends what you mean by "CSP support" (depends what you put into your CSP). Please read above comments; there are still instances of `javascript:''` in the code, including the `xsiframe` linker (but not in the `direct_install` linker), so that depends what you're using and what you put in your CSP.

**Comment by tbroyer** [comment]
https://gwt-review.googlesource.com/c/gwt/+/19620

**Comment by acrestan** [comment]
This issues should not be closed as even version 2.8.2 still uses '_eval_' javascript call. (ie `ClientSerializationStreamReader`)
This forces the use CSP directive 'unsafe-eval'... which is not very good.

[GWT_CSP.txt](https://github.com/gwtproject/gwt/files/1531755/GWT_CSP.txt)

Regards,

**Comment by niloc132** [comment]
Good catch. Here is the code that invokes that eval, and only does so if the rpc stream format version is less than `8`, the current version:

    /*
     * Need to read the version from the string since version is later set in
     * super.prepareToRead() from the evaluated payload but that is too late.
     */
    if (readVersion(encoded) < SERIALIZATION_STREAM_JSON_VERSION) {
      // Versions prior to 8 uses almost JSON with Javascript is special cases; e.g., using ].concat([,
      // non-stringified NaN/Infinity or ''+'' concatenated strings.
      results = eval(encoded);
    } else {
      results = JsonUtils.safeEval(encoded);
    }


On the server, that version is set only if some other part of the stream required an old format:

    /**
     * Notice that the field are written in reverse order that the client can just
     * pop items out of the stream.
     */
    private void writeHeader(LengthConstrainedArray stream) {
      stream.addToken(getFlags());
      if (stream.isJavaScript() && getVersion() >= SERIALIZATION_STREAM_JSON_VERSION) {
        // Ensure we are not using the JSON supported version if stream is Javascript instead of JSON
        stream.addToken(SERIALIZATION_STREAM_JSON_VERSION - 1);
      } else {
        stream.addToken(getVersion());
      }
    }

This is set in two cases - when strings are longer than 64kb (to correctly deal with a hosted mode issue), and if more than 32k values are in the stream (to deal with a bug in IE6/7). I think we can probably safely remove that ie6/7 support (and when I do so, I will test with IE8 to ensure we are correctly compatible).

Hosted mode is a different story: from my (very quick) read of the legacy dev mode version of the ClientSerializationStreamReader, it actually doesn't need that 64kb string limitation any more, and in fact strips out the `+` tokens automatically before parsing the string:

    @Override
    public void prepareToRead(String encoded) throws SerializationException {
      try {
        List<JsStatement> stmts = JsParser.parse(SourceOrigin.UNKNOWN, JsRootScope.INSTANCE,
            new StringReader(encoded));
        ArrayConcatEvaler arrayConcatEvaler = new ArrayConcatEvaler();
        arrayConcatEvaler.acceptList(stmts);
        StringConcatEvaler stringConcatEvaler = new StringConcatEvaler();
        stringConcatEvaler.acceptList(stmts);

I'm going to split this off into its own issue for its own fix/test/verification, since this current one was already closed for 2.8.2, and this is the only case remaining (either huge payloads with more than 32k values or object references, or very large strings) case. If we see more issues, perhaps we'll need to reevaluate.
